src/python/pants/backend/python/tasks2/pytest_run.py

```diff
@@ -485,25 +485,44 @@ def _get_target_from_test(self, test_info, targets):
     file_info = test_info['file']
     return relsrc_to_target.get(file_info)
 
-  def _run_tests(self, targets):
+  def _partition(self, targets):
+    # TODO(John Sirois): Consume `py.test` pexes matched to the partitioning in effect after
+    # https://github.com/pantsbuild/pants/pull/4638 lands.
     if self.get_options().fast:
-      result = self._do_run_tests(targets)
-      if not result.success:
-        raise ErrorWhileTesting(failed_targets=result.failed_targets)
+      return tuple(targets),
     else:
-      results = {}
-      for target in targets:
-        rv = self._do_run_tests([target])
-        results[target] = rv
-        if not rv.success and self.get_options().fail_fast:
-          break
-      for target in sorted(results):
-        self.context.log.info('{0:80}.....{1:>10}'.format(target.id, str(results[target])))
-      failed_targets = [target for target, _rv in results.items() if not _rv.success]
-      if failed_targets:
-        raise ErrorWhileTesting(failed_targets=failed_targets)
+      return tuple((target,) for target in targets)
+
+  def _run_tests(self, targets):
+    partitions = self._partition(targets)
+
+    results = {}
+    for partition in partitions:
+      rv = self._do_run_tests(partition)
+      results[partition] = rv
+      if not rv.success and self.get_options().fail_fast:
+        break
+
+    for partition in sorted(results):
+      rv = str(results[partition])
+      for target in partition:
+        self.context.log.info('{0:80}.....{1:>10}'.format(target.id, rv))
+
+    failed_targets = [target
+                      for _rv in results.values() if not _rv.success
+                      for target in _rv.failed_targets]
+    if failed_targets:
+      raise ErrorWhileTesting(failed_targets=failed_targets)
 
   def _do_run_tests(self, targets):
+    with self.invalidated(targets,
+                          # Re-run tests when the code they test (and depend on) changes.
+                          invalidate_dependents=True) as invalidation_check:
+
+      invalid_tgts = [tgt for vts in invalidation_check.invalid_vts for tgt in vts.targets]
+      return self._run_pytest(invalid_tgts)
```

Comment:
This line does not raise on failure - it always returns a PytestResult. As a result, both test failures and successes are cached. Needs fixing...